lib/models/addon.js

```diff
@@ -167,36 +168,41 @@ function warn(message) {
 let addonProto = {
 
   /**
-  Initializes the addon.  If you override this method make sure and call `this._super.init && this._super.init.apply(this, arguments);` or your addon will not work.
-
-  @public
-  @method init
-  @param {Project|Addon} parent The project or addon that directly depends on this addon
-  @param {Project} project The current project (deprecated)
-
-  @example
-  ```js
-  init(parent, project) {
-    this._super.init && this._super.init.apply(this, arguments);
-    this._someCustomSetup();
-  }
-  ```
-  */
+   * Initializes the addon.  If you override this method make sure and call
+   * `this._super.init && this._super.init.apply(this, arguments);` or your addon will not work.
+   *
+   * @public
+   * @method init
+   * @param {Project|Addon} parent The project or addon that directly depends on this addon
+   * @param {Project} project The current project (deprecated)
+   *
+   * @example
+   * ```js
+   * init(parent, project) {
+   *   this._super.init && this._super.init.apply(this, arguments);
+   *   this._someCustomSetup();
+   * }
+   * ```
+   */
   init(parent, project) {
     this._super();
     this.parent = parent;
     this.project = project;
     this.ui = project && project.ui;
-    this.addonPackages = {};
+    this.addonPackages = Object.create(null);
     this.addons = [];
-    this.addonDiscovery = new AddonDiscovery(this.ui);
     this.addonsFactory = new AddonsFactory(this, this.project);
     this.registry = p.defaultRegistry(this);
     this._didRequiredBuildPackages = false;
 
     if (!this.root) {
       throw new Error('Addon classes must be instantiated with the `root` property');
     }
+
+    if (!this.name) {
```

Comment:
seems like a new behaviour? should we unit test this as well? also, what's the reasoning behind throwing an error?

Reply:
Actually that was already in the class when I started working on it. It might have been farther down in the constructor.